Support md5 hashing part of the cache key above a certain length

matteius opened the pull request:
This change is largely to ensure that the removing of 'KEY_FUNCTION': make_hashed_key which has always prevented us from seeing what the key values actually are in memcached.  There is a limitation that memcached keys be not longer 250 characters and contain no spaces.   This updates the get_cache_key to default to 240 with an override available on the BaseCacheBackend constructor.

The reason for 240 is to give a 10 character space to the default django KEY_FUNCTION which prepends a version and the KEY_PREFIX if its define. 

Also included is a python 3.5 compatibility change for one of the unit tests.  I discovered this while running the tests for nap.

jkomusin commented on nap/cache/base.py:
```diff
@@ -28,16 +34,19 @@ def get_timeout_from_header(self, response):
-
-        cache_key = "%(resource_name)s::%(url)s" % key_parts
+        if (len(cache_key) + len(url)) > self.cache_max_key_size:
+            cache_key += md5(url.encode('utf-8')).hexdigest()
```
Would a `log.debug()` statement be appropriate here to provide some dev visibility into what the original URL was?

matteius replied:
I think that would be wise ... its a tough debate between info and debug for this one because almost none of our apps log debug statements, and yet it would be nice to see these in our logs.  One argument for info is that most cache keys based on a sane API url would be under 240 characters.  Thoughts?

jkomusin replied:
True. Thinking about it more I think info might be more appropriate for the reason you specified. It's useful information to log in a production environment.

matteius commented:
The biggest area for improvement I can see here is that I would like to make the max cache key size easily cutomizeable for projects that aren't using memcached and could support a longer key (or maybe want this feature disabled).   I think in the current form it allows someone to override it on a per resource basis using the Meta: class, but not on a per project basis using a NapConfig.   I tried to search for examples of us overriding anything about nap using a NapConfig but came back empty handed.   Does anyone have some expertise on this area of nap, which is defaulting values and overriding those behaviors?

matteius commented:
I've updated to add the requested logging statement and to set the default log level of INFO.  I also adjusted the conditional to define the setting of MAX_CACHE_KEY_LENGTH = None to be disable length checking.
